src/Debug.h

```diff
@@ -35,7 +37,7 @@
 
 /* context-based debugging, the actual type is subject to change */
 typedef int Ctx;
-Ctx ctx_enter(const char *descr);
+Ctx ctx_enter(const std::string &descr);
```

Comment:
The problem is less the current c-strings and more that the callers are about to not have c-string URLs to pass in. The callers would be adding an allocate+copy in every caller (converting *to* c-string) to avoid this code managing its own memory requirements.

Also, the workaround of having caller release the c-string they allocated for ctx_enter() using xfree() and ctx_exit() together - will lead to memory leaks each time one of these many new exception on parse error happens.

Reply:
FWIW: I looked first at removing the Ctx mechanism entirely. Unfortunately the details its debugging seem to be a useful tool for identifying problems when cbdata callback nesting is involved (eg the doCallouts() stacks). It will probably need to stay around until we prune out the old callback functions properly.

Reply:
> the callers are about to not have c-string URLs to pass in. The callers would be adding an allocate+copy

A PR introducing such a change should probably be rejected for the same reasons why this PR should not add those allocations and copies.

> I looked first at removing the Ctx mechanism entirely. Unfortunately the details its debugging seem to be a useful tool

Yes, context debugging can be very useful. It was added for a good reason. Nevertheless, it is not essential and should be disabled if/when it presents more problems than it solves, until a better implementation is available. As far as I can tell based on your feedback, that time has not come yet, but it may be approaching.

Reply:
Storing in non-char* format requires reallocation to convert to char*. So you are saying that any attempt to make URLs be stored in anything accept char* are going to be rejected because this rarely used debug API uses char*. Understood.